File: clients/javascript/examples/basic.js
/**
 * BaraDB JavaScript Client — Basic Examples
 *
 * Run: node examples/basic.js
 * Requires BaraDB server on localhost:9472.
 */

const { Client, WireValue, QueryBuilder } = require('../baradb');

const HOST = 'localhost';
const PORT = 9472;

async function exampleConnection() {
  console.log('=== Connection ===');
  const client = new Client(HOST, PORT);
  await client.connect();
  console.log(`Connected: ${client.isConnected()}`);
  console.log(`Ping: ${await client.ping()}`);
  await client.close();
  console.log(`Connected after close: ${client.isConnected()}`);
  console.log();
}

async function exampleSimpleQuery() {
  console.log('=== Simple Query ===');
  const client = new Client(HOST, PORT);
  await client.connect();
  const result = await client.query("SELECT 42 as answer, 'BaraDB' as db");
  console.log(`Columns: ${result.columns.join(', ')}`);
  console.log(`Row count: ${result.rowCount}`);
  for (const row of result) {
    console.log(`  answer=${row.answer}, db=${row.db}`);
  }
  await client.close();
  console.log();
}

async function exampleParameterizedQuery() {
  console.log('=== Parameterized Query ===');
  const client = new Client(HOST, PORT);
  await client.connect();
  const result = await client.queryParams(
    'SELECT $1 as num, $2 as txt, $3 as flag',
    [
      WireValue.int64(123),
      WireValue.string('hello world'),
      WireValue.bool(true),
    ]
  );
  for (const row of result) {
    console.log(`  num=${row.num}, txt=${row.txt}, flag=${row.flag}`);
  }
  await client.close();
  console.log();
}

async function exampleQueryBuilder() {
  console.log('=== Query Builder ===');
  const client = new Client(HOST, PORT);
  await client.connect();
  const sql = await new QueryBuilder(client)
    .select('id', 'name')
    .from('users')
    .where('active = true')
    .orderBy('name', 'ASC')
    .limit(5)
    .build();
  console.log(`Generated SQL: ${sql}`);
  await client.close();
  console.log();
}

async function exampleVector() {
  console.log('=== Vector Value ===');
  const client = new Client(HOST, PORT);
  await client.connect();
  const result = await client.queryParams(
    'SELECT $1 as embedding',
    [WireValue.vector([0.1, 0.2, 0.3, 0.4])]
  );
  for (const row of result) {
    console.log(`  embedding type: ${typeof row.embedding}`);
  }
  await client.close();
  console.log();
}

async function exampleDdlDml() {
  console.log('=== DDL & DML ===');
  const client = new Client(HOST, PORT);
  await client.connect();

  try {
    await client.execute('DROP TABLE IF EXISTS demo_products');
  } catch (err) {
    console.log(`Cleanup warning: ${err.message}`);
  }

  await client.execute('CREATE TABLE demo_products (id INT PRIMARY KEY, name STRING, price FLOAT)');
  const affected = await client.execute("INSERT INTO demo_products (id, name, price) VALUES (1, 'Widget', 9.99)");
  console.log(`Insert affected rows: ${affected}`);

  const result = await client.query('SELECT * FROM demo_products');
  console.log(`Select returned ${result.rowCount} row(s)`);
  for (const row of result) {
    console.log(`  ${JSON.stringify(row)}`);
  }

  await client.execute('DROP TABLE demo_products');
  console.log('Table dropped');
  await client.close();
  console.log();
}

async function main() {
  console.log('BaraDB JavaScript Client Examples');
  console.log('Make sure BaraDB is running on localhost:9472');
  console.log();

  const examples = [
    exampleConnection,
    exampleSimpleQuery,
    exampleParameterizedQuery,
    exampleQueryBuilder,
    exampleVector,
    exampleDdlDml,
  ];

  for (const fn of examples) {
    try {
      await fn();
    } catch (err) {
      console.error(`ERROR in ${fn.name}: ${err.message}`);
    }
  }
}

main().catch(console.error);

File: clients/rust/examples/basic.rs
// BaraDB Rust Client — Basic Examples
// Make sure BaraDB is running on localhost:9472.

use baradb::{Client, QueryBuilder, WireValue};

#[tokio::main]
async fn main() {
    println!("BaraDB Rust Client Examples");
    println!("Make sure BaraDB is running on localhost:9472");
    println!();

    if let Err(e) = example_connection().await {
        eprintln!("ERROR: {}", e);
    }

    if let Err(e) = example_simple_query().await {
        eprintln!("ERROR: {}", e);
    }

    if let Err(e) = example_parameterized_query().await {
        eprintln!("ERROR: {}", e);
    }

    if let Err(e) = example_query_builder().await {
        eprintln!("ERROR: {}", e);
    }

    if let Err(e) = example_vector().await {
        eprintln!("ERROR: {}", e);
    }

    if let Err(e) = example_ddl_dml().await {
        eprintln!("ERROR: {}", e);
    }
}

async fn example_connection() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("=== Connection ===");
    let mut client = Client::connect("127.0.0.1", 9472).await?;
    println!("Connected: {}", client.is_connected());
    println!("Ping: {}", client.ping().await?);
    client.close().await;
    println!("Connected after close: {}", client.is_connected());
    println!();
    Ok(())
}

async fn example_simple_query() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("=== Simple Query ===");
    let mut client = Client::connect("127.0.0.1", 9472).await?;
    let result = client.query("SELECT 42 as answer, 'BaraDB' as db").await?;
    println!("Columns: {:?}", result.columns());
    println!("Row count: {}", result.row_count());
    for row in result.rows() {
        println!("  {:?}", row);
    }
    client.close().await;
    println!();
    Ok(())
}

async fn example_parameterized_query() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("=== Parameterized Query ===");
    let mut client = Client::connect("127.0.0.1", 9472).await?;
    let result = client.query_params(
        "SELECT $1 as num, $2 as txt, $3 as flag",
        &[
            WireValue::Int64(123),
            WireValue::String("hello world".to_string()),
            WireValue::Bool(true),
        ],
    ).await?;
    for row in result.rows() {
        println!("  {:?}", row);
    }
    client.close().await;
    println!();
    Ok(())
}

async fn example_query_builder() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("=== Query Builder ===");
    let mut client = Client::connect("127.0.0.1", 9472).await?;
    let sql = QueryBuilder::new(&mut client)
        .select(&["id", "name"])
        .from("users")
        .where_clause("active = true")
        .order_by("name", "ASC")
        .limit(5)
        .build();
    println!("Generated SQL: {}", sql);
    client.close().await;
    println!();
    Ok(())
}

async fn example_vector() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("=== Vector Value ===");
    let mut client = Client::connect("127.0.0.1", 9472).await?;
    let result = client.query_params(
        "SELECT $1 as embedding",
        &[WireValue::Vector(vec![0.1, 0.2, 0.3, 0.4])],
    ).await?;
    for row in result.rows() {
        println!("  {:?}", row);
    }
    client.close().await;
    println!();
    Ok(())
}

async fn example_ddl_dml() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("=== DDL & DML ===");
    let mut client = Client::connect("127.0.0.1", 9472).await?;

    let _ = client.execute("DROP TABLE IF EXISTS demo_products").await;

    client.execute(
        "CREATE TABLE demo_products (id INT PRIMARY KEY, name STRING, price FLOAT)",
    ).await?;
    let affected = client.execute(
        "INSERT INTO demo_products (id, name, price) VALUES (1, 'Widget', 9.99)",
    ).await?;
    println!("Insert affected rows: {}", affected);

    let result = client.query("SELECT * FROM demo_products").await?;
    println!("Select returned {} row(s)", result.row_count());
    for row in result.rows() {
        println!("  {:?}", row);
    }

    client.execute("DROP TABLE demo_products").await?;
    println!("Table dropped");
    client.close().await;
    println!();
    Ok(())
}